grudge/tools.py

```diff
@@ -235,4 +239,105 @@ def build_jacobian(actx, f, base_state, stepsize):
     return mat
 
 
+def map_subarrays(
+        f: Callable[[np.ndarray], np.ndarray],
+        in_shape: Tuple[int, ...], out_shape: Tuple[int, ...],
+        ary: np.ndarray, *, return_nested=False) -> np.ndarray:
+    """
+    Apply a function *f* to subarrrays of shape *in_shape* of an
+    :class:`numpy.ndarray`, typically (but not necessarily) of dtype
+    :class:`object`. Return an :class:`numpy.ndarray` of the same dtype,
+    with the corresponding subarrays replaced by the return values of *f*,
+    and with the shape adapted to reflect *out_shape*.
+
+    Similar to :class:`numpy.vectorize`.
+
+    *Example 1:* given a function *f* that maps arrays of shape ``(2, 2)`` to scalars
+    and an input array *ary* of shape ``(3, 2, 2)``, the call::
+
+        map_subarrays(f, (2, 2), (), ary)
+
+    will produce an array of shape ``(3,)`` containing the results of calling *f* on
+    the 3 subarrays of shape ``(2, 2)`` in *ary*.
+
+    *Example 2:* given a function *f* that maps arrays of shape ``(2,)`` to arrays of
+    shape ``(2, 2)`` and an input array *ary* of shape ``(3, 2)``, the call::
+
+        map_subarrays(f, (2,), (2, 2), ary)
+
+    will produce an array of shape ``(3, 2, 2)`` containing the results of calling
+    *f* on the 3 subarrays of shape ``(2,)`` in *ary*. The call::
+
+        map_subarrays(f, (2,), (2, 2), ary, return_nested=True)
+
+    will instead produce an array of shape ``(3,)`` with each entry containing an
+    array of shape ``(2, 2)``.
+
+    :param f: the function to be called.
+    :param in_shape: the shape of any inputs to *f*.
+    :param out_shape: the shape of the result of calling *f* on an array of shape
+        *in_shape*.
+    :param ary: a :class:`numpy.ndarray` instance.
+    :param return_nested: if *out_shape* is nontrivial, this flag indicates whether
+        to return a nested array (containing one entry for each application of *f*),
+        or to return a single, higher-dimensional array.
+
+    :returns: an array with the subarrays of shape *in_shape* replaced with
+        subarrays of shape *out_shape* resulting from the application of *f*.
+    """
+    if not isinstance(ary, np.ndarray):
+        if len(in_shape) != 0:
+            raise ValueError(f"found scalar, expected array of shape {in_shape}")
+        return f(ary)
+    else:
+        if (
+                ary.ndim < len(in_shape)
+                or ary.shape[ary.ndim-len(in_shape):] != in_shape):
+            raise ValueError(
+                f"array of shape {ary.shape} is incompatible with function "
+                f"expecting input shape {in_shape}")
+        base_shape = ary.shape[:ary.ndim-len(in_shape)]
+        if len(base_shape) == 0:
+            return f(ary)
+        else:
+            in_slice = tuple(slice(0, n) for n in in_shape)
+            out_slice = tuple(slice(0, n) for n in out_shape)
+            if return_nested:
+                result = np.empty(base_shape, dtype=ary.dtype)
```

Comment:
```suggestion
                result = np.empty(base_shape, dtype=object)
```
I think this one at least needs to be an object array, since the entries going into it may be arrays.

Reply:
I made it deduce the `dtype` from the function application results.